File: implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/newsletter/JooqNewsletterRepoTest.java
package ch.difty.scipamato.core.persistence.newsletter;

import static ch.difty.scipamato.core.db.tables.Newsletter.NEWSLETTER;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.sql.Timestamp;

import org.jooq.TableField;
import org.junit.Test;
import org.mockito.Mock;

import ch.difty.scipamato.common.persistence.paging.PaginationContext;
import ch.difty.scipamato.core.db.tables.records.NewsletterRecord;
import ch.difty.scipamato.core.entity.newsletter.Newsletter;
import ch.difty.scipamato.core.entity.newsletter.NewsletterFilter;
import ch.difty.scipamato.core.persistence.EntityRepository;
import ch.difty.scipamato.core.persistence.JooqEntityRepoTest;

@SuppressWarnings("ResultOfMethodCallIgnored")
public class JooqNewsletterRepoTest extends
    JooqEntityRepoTest<NewsletterRecord, Newsletter, Integer, ch.difty.scipamato.core.db.tables.Newsletter, NewsletterRecordMapper, NewsletterFilter> {

    private static final Integer SAMPLE_ID = 2;

    private JooqNewsletterRepo repo;

    @Mock
    private Newsletter             unpersistedEntity;
    @Mock
    private Newsletter             persistedEntity;
    @Mock
    private NewsletterRecord       persistedRecord;
    @Mock
    private NewsletterRecordMapper mapperMock;
    @Mock
    private NewsletterFilter       filterMock;
    @Mock
    private Newsletter             newsletterMock;
    @Mock
    private PaginationContext      pageableMock;

    @Override
    protected void testSpecificSetUp() {
        when(unpersistedEntity.getVersion()).thenReturn(0);
        when(persistedEntity.getVersion()).thenReturn(0);
        when(persistedRecord.getVersion()).thenReturn(0);
    }

    @Override
    protected EntityRepository<Newsletter, Integer, NewsletterFilter> makeRepoSavingReturning(
        final NewsletterRecord returning) {
        return new JooqNewsletterRepo(getDsl(), getMapper(), getSortMapper(), getFilterConditionMapper(),
            getDateTimeService(), getInsertSetStepSetter(), getUpdateSetStepSetter(), getApplicationProperties()) {
            @Override
            protected NewsletterRecord doSave(final Newsletter entity, final String languageCode) {
                return returning;
            }
        };
    }

    @Override
    protected Integer getSampleId() {
        return SAMPLE_ID;
    }

    @Override
    protected EntityRepository<Newsletter, Integer, NewsletterFilter> getRepo() {
        if (repo == null) {
            repo = new JooqNewsletterRepo(getDsl(), getMapper(), getSortMapper(), getFilterConditionMapper(),
                getDateTimeService(), getInsertSetStepSetter(), getUpdateSetStepSetter(), getApplicationProperties());
        }
        return repo;
    }

    @Override
    protected EntityRepository<Newsletter, Integer, NewsletterFilter> makeRepoFindingEntityById(
        final Newsletter entity) {
        return new JooqNewsletterRepo(getDsl(), getMapper(), getSortMapper(), getFilterConditionMapper(),
            getDateTimeService(), getInsertSetStepSetter(), getUpdateSetStepSetter(), getApplicationProperties()) {

            @Override
            public Newsletter findById(Integer id, int version) {
                return entity;
            }
        };
    }

    @Override
    protected Newsletter getPersistedEntity() {
        return persistedEntity;
    }

    @Override
    protected Newsletter getUnpersistedEntity() {
        return unpersistedEntity;
    }

    @Override
    protected NewsletterRecord getPersistedRecord() {
        return persistedRecord;
    }

    @Override
    protected NewsletterRecordMapper getMapper() {
        return mapperMock;
    }

    @Override
    protected ch.difty.scipamato.core.db.tables.Newsletter getTable() {
        return NEWSLETTER;
    }

    @Override
    protected TableField<NewsletterRecord, Integer> getTableId() {
        return NEWSLETTER.ID;
    }

    @Override
    protected void expectEntityIdsWithValues() {
        when(unpersistedEntity.getId()).thenReturn(SAMPLE_ID);
        when(unpersistedEntity.getVersion()).thenReturn(0);
        when(persistedRecord.getId()).thenReturn(SAMPLE_ID);
        when(persistedRecord.getVersion()).thenReturn(1);
    }

    @Override
    protected void expectUnpersistedEntityIdNull() {
        when(unpersistedEntity.getId()).thenReturn(null);
        when(unpersistedEntity.getVersion()).thenReturn(0);
    }

    @Override
    protected void verifyUnpersistedEntityId() {
        verify(getUnpersistedEntity()).getId();
    }

    @Override
    protected void verifyPersistedRecordId() {
        verify(persistedRecord).getId();
    }

    @Override
    protected NewsletterFilter getFilter() {
        return filterMock;
    }

    @Override
    protected TableField<NewsletterRecord, Integer> getRecordVersion() {
        return NEWSLETTER.VERSION;
    }

    @Test
    public void mergingPaperIntoNewsletter_withInsertAttemptNotSucceeding_returnsEmptyOptional() {
        JooqNewsletterRepo repo = new JooqNewsletterRepo(getDsl(), getMapper(), getSortMapper(),
            getFilterConditionMapper(), getDateTimeService(), getInsertSetStepSetter(), getUpdateSetStepSetter(),
            getApplicationProperties()) {
            @Override
            int tryInserting(final int newsletterId, final long paperId, final Integer newsletterTopicId,
                final Timestamp ts) {
                return 0;
            }
        };
        assertThat(repo.mergePaperIntoNewsletter(1, 2L, 3, "en")).isNotPresent();
    }
}

File: core/core-persistence-jooq/src/test/kotlin/ch/difty/scipamato/core/persistence/newsletter/JooqNewsletterRepoTest.kt
package ch.difty.scipamato.core.persistence.newsletter

import ch.difty.scipamato.core.db.tables.Newsletter.NEWSLETTER
import ch.difty.scipamato.core.db.tables.records.NewsletterRecord
import ch.difty.scipamato.core.entity.newsletter.Newsletter
import ch.difty.scipamato.core.entity.newsletter.NewsletterFilter
import ch.difty.scipamato.core.persistence.EntityRepository
import ch.difty.scipamato.core.persistence.JooqEntityRepoTest
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.amshove.kluent.shouldBeFalse
import org.jooq.TableField
import org.junit.jupiter.api.Test
import java.sql.Timestamp

internal class JooqNewsletterRepoTest : JooqEntityRepoTest<NewsletterRecord, Newsletter, Int,
    ch.difty.scipamato.core.db.tables.Newsletter, NewsletterRecordMapper, NewsletterFilter>() {

    override val unpersistedEntity = mockk<Newsletter>()
    override val persistedEntity = mockk<Newsletter>()
    override val persistedRecord = mockk<NewsletterRecord>()
    override val unpersistedRecord = mockk<NewsletterRecord>()
    override val mapper = mockk<NewsletterRecordMapper>()
    override val filter = mockk<NewsletterFilter>()
    override val repo: JooqNewsletterRepo = JooqNewsletterRepo(
        dsl,
        mapper,
        sortMapper,
        filterConditionMapper,
        dateTimeService,
        insertSetStepSetter,
        updateSetStepSetter,
        applicationProperties
    )

    override fun makeRepoSavingReturning(returning: NewsletterRecord) =
        object : JooqNewsletterRepo(
            dsl,
            mapper,
            sortMapper,
            filterConditionMapper,
            dateTimeService,
            insertSetStepSetter,
            updateSetStepSetter,
            applicationProperties
        ) {
            override fun doSave(entity: Newsletter, languageCode: String) = returning
        }

    override val sampleId: Int = SAMPLE_ID
    override val table: ch.difty.scipamato.core.db.tables.Newsletter = NEWSLETTER
    override val tableId: TableField<NewsletterRecord, Int> = NEWSLETTER.ID

    override fun makeRepoFindingEntityById(entity: Newsletter): EntityRepository<Newsletter, Int, NewsletterFilter> =
        object : JooqNewsletterRepo(
            dsl,
            mapper,
            sortMapper,
            filterConditionMapper,
            dateTimeService,
            insertSetStepSetter,
            updateSetStepSetter,
            applicationProperties
        ) {
            override fun findById(id: Int, version: Int): Newsletter? = entity
        }

    override fun expectEntityIdsWithValues() {
        every { unpersistedEntity.id } returns SAMPLE_ID
        every { unpersistedEntity.version } returns 0
        every { persistedRecord.id } returns SAMPLE_ID
        every { persistedRecord.version } returns 1
    }

    override fun expectUnpersistedEntityIdNull() {
        every { unpersistedEntity.id } returns null
    }

    override fun verifyUnpersistedEntityId() {
        verify { unpersistedEntity.id }
    }

    override fun verifyPersistedRecordId() {
        verify { persistedRecord.id }
    }

    override val recordVersion: TableField<NewsletterRecord, Int> = NEWSLETTER.VERSION

    @Test
    fun mergingPaperIntoNewsletter_withInsertAttemptNotSucceeding_returnsEmptyOptional() {
        val repo = object : JooqNewsletterRepo(
            dsl,
            mapper,
            sortMapper,
            filterConditionMapper,
            dateTimeService,
            insertSetStepSetter,
            updateSetStepSetter,
            applicationProperties
        ) {
            override fun tryInserting(
                newsletterId: Int,
                paperId: Long,
                newsletterTopicId: Int?,
                ts: Timestamp?
            ): Int = 0
        }
        repo.mergePaperIntoNewsletter(1, 2L, 3, "en").isPresent.shouldBeFalse()
    }

    @Test
    fun handlingInsertedNewsletter_withZeroCount_returnsEmptyOptional() {
        repo.handleInsertedNewsletter(0, 1, 2, "de").isPresent.shouldBeFalse()
    }

    @Test
    fun test() {
        val repo = object : JooqNewsletterRepo(
            dsl,
            mapper,
            sortMapper,
            filterConditionMapper,
            dateTimeService,
            insertSetStepSetter,
            updateSetStepSetter,
            applicationProperties
        ) {
            override fun fetchMergedNewsletter(newsletterId: Int, paperId: Long, languageCode: String) = null
        }
        repo.handleInsertedNewsletter(1, 1, 2, "de").isPresent.shouldBeFalse()
    }

    companion object {
        private const val SAMPLE_ID = 2
    }
}
